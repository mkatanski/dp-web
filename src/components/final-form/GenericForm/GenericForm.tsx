import React from "react";
import { Form } from "react-final-form";

export type GenericFormProps = {
  onSubmit: (data: unknown) => void;
  initialValues: Partial<unknown>;
  children: React.ReactNode;
};

export const GenericForm: React.FC<GenericFormProps> = ({
  onSubmit,
  initialValues,
  children
}: GenericFormProps) => {
  return (
    <Form
      onSubmit={onSubmit}
      initialValues={initialValues}
      subscription={{ submitting: true, pristine: true }}
      render={({ handleSubmit }) => {
        return <form onSubmit={handleSubmit}>{children}</form>;
      }}
    />
  );
};
