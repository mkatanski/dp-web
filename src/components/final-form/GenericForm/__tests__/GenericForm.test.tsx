/* eslint-disable react/display-name */
import React from "react";
import TestRenderer from "react-test-renderer";
import "jest-styled-components";
import { Field } from "react-final-form";
import { render, fireEvent, wait } from "@testing-library/react";

import { GenericForm } from "../GenericForm";

describe("components", () => {
  describe("GenericForm", () => {
    const submitHandler: jest.Mock = jest.fn();
    it("should render GenericForm component with Field component", () => {
      const tree = TestRenderer.create(
        <GenericForm
          onSubmit={submitHandler}
          initialValues={{ nickname: "MirasHandlarz" }}
        >
          <Field
            name="nickname"
            component="input"
            type="text"
            placeholder="Nickname"
          />
        </GenericForm>
      );
      expect(tree).toMatchSnapshot();
    });

    it("should submit correct data", async () => {
      const submitHandler = jest.fn();

      const { getByTestId } = render(
        <GenericForm
          onSubmit={submitHandler}
          initialValues={{ nickname: "MirasHandlarz" }}
        >
          <Field
            data-testid="input"
            name="nickname"
            component="input"
            type="text"
            placeholder="Nickname"
          />
          <button role="submit" data-testid="submit" type="submit">
            submitForm
          </button>
        </GenericForm>
      );
      await wait(() => {
        fireEvent.change(getByTestId("input"), {
          target: { value: "blytEsT123$bly" }
        });
      });

      await wait(() => {
        fireEvent.click(getByTestId("submit"));
      });

      expect(submitHandler).toBeCalledTimes(1);
      expect(submitHandler.mock.calls[0][0]).toEqual({
        nickname: "blytEsT123$bly"
      });
    });
  });
});
