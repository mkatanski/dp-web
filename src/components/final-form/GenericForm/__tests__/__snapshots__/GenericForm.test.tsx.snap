// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components GenericForm should render GenericForm component with Field component 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    name="nickname"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Nickname"
    type="text"
    value="MirasHandlarz"
  />
</form>
`;
